rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,302 @@
+up disk space from unused binaries.
+
+To enable concurrent operation of client tools, a locking mechanisms utilizing
+[syscall.Flock](https://pkg.go.dev/syscall#Flock) (for Linux and macOS) and
+[LockFileEx](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex)
+(for Windows) will be used.
+
+A side effect of automatic updates will be users will be able to connect to
+different Teleport clusters (running different versions of Teleport, including
+pre-releases) without juggling multiple versions of client tools.
+
+```
+$ tree ~/.tsh
+~/.tsh
+├── current-profile
+├── keys
+│  └── proxy.example.com
+│     ├── cas
+│     │  └── example.com.pem
+│     ├── certs.pem
+│     ├── foo
+│     ├── foo-ssh
+│     │  └── example.com-cert.pub
+│     ├── foo-x509.pem
+│     └── foo.pub
+├── known_hosts
+├── proxy.example.com.yaml
+└── versions
+   ├── 15.0.0
+   │  └── bin
+   │     ├── tctl
+   │     └── tsh
+   ├── 15.1.1
+   │  └── bin
+   │     ├── tctl
+   │     └── tsh
+   └── versions.yaml
+```
+
+Users can cancel client tools updates using `Ctrl-C`. This may be needed if the
+user is on a low bandwidth connection (LTE or public Wi-Fi), if the Teleport
+download server is inaccessible, or the user urgently needs to access the
+cluster and can not wait for the update to occur.
+
+```
+$ tsh login --proxy=proxy.example.com
+Client tools are out of date, updating to vX.Y.Z.
+Update progress: [▒▒▒▒▒▒     ] (Ctrl-C to cancel update)
+
+[...]
+```
+
+An environment variable `TELEPORT_TOOLS_VERSION` will be introduced that can be
+`X.Y.Z` (use specific version) or `off` (do not update). This environment
+variable can be used as a emergency workaround for a known issue, pinning to a
+specific version in CI/CD, or for debugging.
+
+During re-execution, child process will inherit all environment variables and
+flags. `TELEPORT_TOOLS_VERSION=off` will be added during re-execution to
+prevent infinite loops.
+
+Automatic updates will not be used if `tctl` is connecting to the Auth Service
+over localhost.
+
+##### Errors and warnings
+
+If cluster administrator has chosen not to enroll client tools in automatic
+updates and does not self-manage client tools updates as outlined in
+[Self-managed client tools updates](#self-managed-client-tools-updates), a
+series of warnings and errors with increasing urgency will be shown to the
+user.
+
+If the version of client tools is within the same major version as advertised
+by the cluster, a warning will be shown to urge the user to enroll in automatic
+updates. Warnings will not prevent the user from using client tools that are
+slightly out of date.
+
+```
+$ tsh login --proxy=proxy.example.com
+Warning: Client tools are out of date, update to vX.Y.Z.
+
+Update Teleport to vX.Y.Z from https://goteleport.com/download or your system
+package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
```

Comment:
What will `tsh autoupdate enroll` command actually do if upgrades are managed server-side? I couldn't find this in the doc.

Reply:
Forgot to update that section when I moved enrollment fully server side.

I'll update the docs links when I push the docs.